test/Microsoft.TemplateEngine.Edge.UnitTests/SdkVersionConstraintTests.cs

```diff
@@ -0,0 +1,163 @@
+            A.CallTo(() => settings.Components.OfType<ITemplateConstraintFactory>()).Returns(new[] { new SdkVersionConstraintFactory() });
+
+            var constraintManager = new TemplateConstraintManager(settings);
+
+            //Workaround needed
+            //A.CallTo(() => sdkInfoProvider.GetVersionAsync(A<CancellationToken>._)).Returns(t);
+
+            var evaluateResult = constraintManager.EvaluateConstraintAsync(configModel.Constraints.Single().Type, configModel.Constraints.Single().Args, default).Result;
+            Assert.Equal(allowed ? TemplateConstraintResult.Status.Allowed : TemplateConstraintResult.Status.Restricted, evaluateResult.EvaluationStatus);
+        }
+
+        [Theory]
+        [InlineData("1.1.1", new[] { "0.1.2", "1.2.3", "3.4.5" }, true)]
+        [InlineData("1.1.1", new[] { "0.1.2", "1.2.2", "3.4.5" }, false)]
+        [InlineData("1.1.1", new[] { "0.1.2", "1.2.3", "4.5.6" }, true)]
+        public async Task Evaluate_AlternativeInstalledVersions(string sdkVersion, IReadOnlyList<string> installedVersions, bool hasAlternativeInstalled)
+        {
+            var config = new
+            {
+                identity = "test-constraint-01",
+                constraints = new
+                {
+                    specVersions = new
+                    {
+                        type = "sdk-version",
+                        args = new[] { "1.2.3-*", "4.5.*" }
+                    }
+                }
+            };
+
+            var configModel = SimpleConfigModel.FromJObject(JObject.FromObject(config));
+            ISdkInfoProvider sdkInfoProvider = new SdkInfoProviderMock(sdkVersion, installedVersions); //A.Fake<ISdkInfoProvider>();
+            IEngineEnvironmentSettings settings = A.Fake<IEngineEnvironmentSettings>();
+            A.CallTo(() => settings.Components.OfType<ISdkInfoProvider>()).Returns(new[] { sdkInfoProvider });
+            A.CallTo(() => settings.Components.OfType<ITemplateConstraintFactory>()).Returns(new[] { new SdkVersionConstraintFactory() });
+
+            var constraintManager = new TemplateConstraintManager(settings);
+
+            var evaluateResult = await constraintManager.EvaluateConstraintAsync(configModel.Constraints.Single().Type, configModel.Constraints.Single().Args, default).ConfigureAwait(false);
+            Assert.Equal(TemplateConstraintResult.Status.Restricted, evaluateResult.EvaluationStatus);
+            Assert.StartsWith(
+                hasAlternativeInstalled
+                    ? "Sample CTA with alternatives"
+                    : "Sample CTA without alternatives",
+                evaluateResult.CallToAction);
+        }
+
+        // This is a workaround in a weird bug with FakeItEasy, when using:
+        //   ISdkInfoProvider sdkInfoProvider = A.Fake<ISdkInfoProvider>();
+        //   A.CallTo(() => sdkInfoProvider.GetVersionAsync(A<CancellationToken>._)).Returns(Task.FromResult(sdkVersion));
```

Comment:
Did you also tried `A.CallTo(() => sdkInfoProvider.GetVersionAsync(default)).Returns(sdkVersion)`
https://fakeiteasy.readthedocs.io/en/stable/faking-async-methods/
I think the issue might be in trying to "fake" cancellation token, but badly reported; I'm not sure if faking structs is supported.

Reply:
Yes - unfortunately didn't help. `CancellationToken` is not actually being faked here - the `Task` is. In some random cases the Task was returning empty string within the tested method (not happening if I create mock manually)